### PR TITLE
feat: deferred rendering pipeline, memory optimization, and async map cleanup

### PR DESCRIPTION
**Pull request by vllsystems**
## Summary  
  
Overhaul the GL rendering pipeline inspired by OTClient's Redemption DrawPool architecture.  
Drawing commands are now accumulated into a deferred command list, merged by  
state (texture + blend mode), and flushed in minimal draw calls. Sprite sheet  
pixel data is released after GPU upload to reduce RAM usage, and map tab  
closing is performed asynchronously to prevent UI freezes on large maps.  
  
### Rendering pipeline (Fase 1–3)  
- **DrawPool + deferred rendering:** all draw calls (`drawTexturedQuad`,  
  `drawColoredQuad`, `drawPolygon`, `drawTriangleFan`, `drawRoundedRect`,  
  `drawThickLineSegment`) now push `DrawCommand` structs to a command list  
  instead of issuing immediate GL calls.  
- **State merging:** consecutive commands with the same state (texture ID +  
  blend mode) are merged into a single batch, reducing draw calls from  
  ~100–500 to ~10–30 per frame.  
- **White pixel texture:** a 1×1 white texture replaces the `uUseTexture`  
  uniform, allowing colored geometry and textured geometry to share the same  
  shader path and merge across batches.  
- **Blend mode in DrawState:** blend mode changes (used by `LightDrawer`) are  
  tracked per command instead of relying on external `glBlendFunc` calls.  
- **Element Buffer Object (EBO):** quads use 4 vertices + 6 indices instead of  
  6 duplicated vertices, reducing vertex memory by ~33%.  
- **Buffer streaming:** ring buffer with `glMapBufferRange` and orphaning  
  reduces CPU↔GPU stalls on large frames.  
  
### FBO scene caching (Fase 4)  
- The map scene is rendered to an off-screen FBO and reused on subsequent  
  frames when nothing has changed (no scroll, zoom, floor change, or edit).  
- Overlays (grid, brush, selection box, tooltips, performance stats) render  
  directly on top of the cached FBO.  
- `MapCanvas::Refresh()` calls `markDirty()` to invalidate the cache on edits.  
- `GUI::DoUndo()` and `GUI::DoRedo()` now call `RefreshView()` instead of  
  `root->Refresh()` to properly propagate dirty state.  
  
### Memory optimization  
- Sprite sheet pixel data (`sheet->data`) is released after GL texture upload.  
  If the data is needed again (e.g., for `getSprite()` or after GC eviction),  
  it is reloaded from disk on demand.  
- Sprite cache (`sprites` map) is cleared in `SpriteAppearances::unload()`.  
- `exportSheetImage()` guards against null data.  
  
### Async map cleanup  
- `MapTab::~MapTab()` now deletes the `Editor` (and its map, tiles, items,  
  action queue) on a detached background thread, preventing UI freezes when  
  closing large maps.  
- On Linux, `malloc_trim(0)` is called after cleanup to return freed memory  
  to the OS.  
  
## Files changed  
  
- `source/gl_renderer.h`  
- `source/gl_renderer.cpp`  
- `source/light_drawer.h`  
- `source/light_drawer.cpp`  
- `source/map_drawer.h`  
- `source/map_drawer.cpp`  
- `source/map_display.cpp`  
- `source/map_tab.cpp`  
- `source/editor.cpp`  
- `source/sprite_appearances.h`  
- `source/sprite_appearances.cpp`  
- `source/gui.cpp`  
  
## HOW TO TEST  
  
1. **Basic rendering** — Open any map. All sprites, items, creatures, and  
   ground tiles should render identically to before.  
2. **Colored overlays** — Select tiles, use brushes, enable grid (`View →  
   Grid`). Colors, selection boxes, and brush indicators should look correct.  
3. **Lights** — Toggle lights with `Shift+L`. The lighting overlay should  
   render correctly (no black screen, no artifacts).  
4. **Zoom and scroll** — Zoom in/out and scroll around the map. The FBO cache  
   should invalidate and re-render on every scroll/zoom change.  
5. **Floor change** — Switch floors. The scene should re-render for the new  
   floor.  
6. **Editing** — Paint tiles, place/remove items, move items. Changes should  
   appear immediately (FBO invalidated via `markDirty()`).  
7. **Undo/Redo** — `Ctrl+Z` / `Ctrl+Y` should update the map correctly.  
8. **Static scene** — Stop interacting with the map for a few seconds. The FBO  
   should be reused without re-rendering (verify via performance stats if  
   available).  
9. **Tooltips and UI** — Hover over items to see tooltips. Rounded rects and  
   text should render correctly.  
10. **Memory usage** — Open a large map, note RAM usage. Close the tab and  
    open a smaller map. RAM should drop (not accumulate).  
11. **Tab close speed** — Close a tab with a large map. The editor should  
    remain responsive (no freeze / "not responding" dialog).  
12. **Palette/item browser** — Browse items in the palette. Sprite icons should  
    load correctly (pixel data reloaded from disk on demand).  
13. **Window resize** — Resize the editor window. The FBO should be recreated  
    at the new resolution without artifacts.

## Summary by CodeRabbit

* **New Features**
  * Scene caching with render-to-texture and blitting for faster map redraws
  * Explicit blend-mode controls and FBO-backed render-to-texture workflow

* **Bug Fixes**
  * More robust sprite-sheet reload and texture release behavior
  * Safer editor/tab shutdown with ensured live-server closure and deferred deletion

* **Performance**
  * Reworked GPU batching/streaming and command-based flushing for fewer draw calls and efficient uploads
  * Smarter scene invalidation to avoid unnecessary re-renders

* **UX**
  * Undo/redo now refreshes editor views more reliably

### CODE REVIEW

File: source/map_tab.cpp [hunk 1]
```diff
@@ -53,8 +57,20 @@ MapTab::MapTab(const MapTab* other) :
...
+		}
+
+		std::thread([ed]() {
+			delete ed;
+#ifdef __linux__
+			malloc_trim(0);
+#endif
+		}).detach();
 	}
 }
```
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_tab.cpp:75]: this is a known trade-off we're keeping intentionally. The editor pointer is nulled out on iref before the detached thread is spawned (line 61), and CloseLiveServer() is called synchronously while the editor is still fully alive (line 65). The wx destruction chain for the tab, canvas, and base MapWindow runs synchronously on the main thread and completes before any deferred event could reference the editor through this widget hierarchy — by the time ~MapTab() returns, the tab has already been removed from the notebook by the AUI manager, so no paint, palette, or dialog callback will route back through it. The detached worker only touches Editor/Map internals that have no dependency on the GUI layer. We agree this pattern is not ideal long-term, but the current sequencing is safe in practice and a proper refactor (moving to a joined worker pool or CallAfter-based deferred deletion) would require broader architectural changes that are out of scope for this PR.

File: source/map_tab.cpp [hunk 2]
```diff
@@ -53,8 +57,20 @@ MapTab::MapTab(const MapTab* other) :
...
+		if (ed->IsLive()) {
+			ed->CloseLiveServer();
+		}
+
+		std::thread([ed]() {
+			delete ed;
+#ifdef __linux__
+			malloc_trim(0);
+#endif
+		}).detach();
```
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_tab.cpp:73]: Fair point in principle, but low-risk in practice. The delete ed path only tears down Editor and Map data — tile arrays, item vectors, undo history — none of which reference GUI singletons, the sprite cache, or any global whose lifetime is tied to wxApp shutdown. The heavy allocations are the entire reason deletion is offloaded to a worker in the first place (it can take hundreds of milliseconds for large maps and would otherwise freeze the UI). On Linux the malloc_trim(0) call is the only post-delete side effect, and it is safe to invoke concurrently. Introducing a std::jthread registry or a drain-on-exit pool is a worthwhile improvement, but it belongs in a dedicated follow-up — not in a rendering-focused PR. We'll track it separately.

File: source/gl_renderer.cpp [hunk 3]
```diff
@@ -746,3 +841,69 @@ void GLRenderer::invalidateTexture(GLuint id) {
...
+	if (fboData.fbo == 0) {
+		return;
+	}
+	glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+	glEnable(GL_BLEND);
+	drawTexturedQuad(0, 0, w, h, fboData.texture, { 255, 255, 255, 255 }, 0.f, 1.f, 1.f, 0.f);
+	flush();
+}
```
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/gl_renderer.cpp:926]: This is cosmetically inconsistent but functionally correct — no change needed. The GL_DEPTH_BUFFER_BIT flag is a no-op in effect: the FBO has no depth attachment, and blitFBO overwrites the entire default-framebuffer viewport, so any prior depth content is irrelevant. The direct glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA) is immediately superseded by the command pipeline inside flush() → flushCommands(), which unconditionally sets glBlendFunc to the same values when activeBlendSrc is 0 (line 790). The activeBlendSrc/activeBlendDst tracking fields are never read outside the command pipeline, so they don't go stale in any observable way. Routing through setBlendMode/resetBlendMode here would add ceremony without changing behaviour.

File: source/map_tab.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_tab.cpp:75]: Glad the sequencing rationale checks out. We'll revisit this when we tackle the broader thread-management refactor. Closing for now.

@coderabbitai resolve

File: source/map_tab.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_tab.cpp:73]: Thanks for acknowledging. We'll track the std::jthread registry / drain-on-exit improvement separately — it's a sound idea but orthogonal to this PR's scope. No changes needed here.

@coderabbitai resolve

File: source/gl_renderer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/gl_renderer.cpp:926]: Correct — purely cosmetic inconsistency with no observable side effects. Keeping as-is to avoid unnecessary churn.

@coderabbitai resolve

File: source/map_drawer.cpp [hunk 4]
```diff
@@ -215,20 +215,78 @@ void MapDrawer::SetupGL() {
...
+		return true;
+	}
+	if (dragging || dragging_draw) {
+		return true;
+	}
+	if (options.show_preview && zoom <= 2.0f) {
+		return true;
+	}
+	return false;
+}
```
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_drawer.cpp:257]: The paste/doodad preview is inherently tile-based — DrawSecondaryMap() positions it using Position(mouse_map_x, mouse_map_y, floor), which are tile coordinates, not pixel coordinates. Moving the mouse within the same tile produces no visual change in the preview, so the current behavior of only calling Refresh() on tile boundary crossings (line 487: map_update && isPasting()) is correct by design. The FBO cache preserves the preview at its last valid tile position between dirty frames, which is the expected result. The AnimationTimer serves animation purposes (show_preview, position indicator, perf stats) — it doesn't mask a missing refresh. EnterDrawingMode() already calls Refresh() (line 2625). No changes needed. Resolve this.

[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_drawer.cpp:257]: I'm the PR author.

[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_drawer.cpp:257]: I've the Member role. Wtf?

**Comment by vllsystems** [comment]
@coderabbit
resolve

File: source/map_drawer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (vllsystems) [reply source/map_drawer.cpp:257]: @coderabbit 
Resolve

**Review by majestyotbr — APPROVED** [review]
